plugins/performance-lab/includes/site-health/far-future-headers/helper.php

```diff
@@ -0,0 +1,320 @@
+	$results = perflab_ffh_check_assets( $assets );
+
+	if ( 'good' !== $results['final_status'] ) {
+		$result['status'] = $results['final_status'];
+		$result['label']  = __( 'Your site does not serve static assets with an effective caching strategy', 'performance-lab' );
+
+		if ( count( $results['details'] ) > 0 ) {
+			$result['actions'] = sprintf(
+				'<p>%s</p>%s<p>%s</p>',
+				esc_html__( 'The following file types do not have the recommended far-future headers. Consider adding or adjusting Cache-Control or Expires headers for these asset types.', 'performance-lab' ),
+				perflab_ffh_get_extensions_table( $results['details'] ),
+				esc_html__( 'Note: "Conditionally cached" means that the browser can re-validate the resource using ETag or Last-Modified headers. This results in fewer full downloads but still requires the browser to make requests, unlike far-future expiration headers that allow the browser to fully rely on its local cache for a longer duration.', 'performance-lab' )
+			);
+		}
+		$result['actions'] .= sprintf(
+			'<p>%s</p>',
+			esc_html__( 'Far-future Cache-Control or Expires headers can be added or adjusted with a small configuration change by your hosting provider.', 'performance-lab' )
+		);
+	}
+
+	return $result;
+}
+
+/**
+ * Checks if far-future expiration headers are enabled for a list of assets.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param  string[] $assets List of asset URLs to check.
+ * @return array{final_status: string, details: array{filename: string, reason: string}[]} Final status and details.
+ */
+function perflab_ffh_check_assets( array $assets ): array {
+	$final_status = 'good';
+	$fail_details = array(); // Array of arrays with 'filename' and 'reason'.
+
+	foreach ( $assets as $asset ) {
+		$response = wp_remote_get( $asset, array( 'sslverify' => false ) );
+
+		// Extract filename from the URL.
+		$path_info = pathinfo( (string) wp_parse_url( $asset, PHP_URL_PATH ) );
+		$filename  = $path_info['basename'] ?? basename( $asset );
+
+		if ( is_wp_error( $response ) ) {
+			// Can't determine headers if request failed, consider it a fail.
+			$final_status   = 'recommended';
+			$fail_details[] = array(
+				'filename' => $filename,
+				'reason'   => __( 'Could not retrieve headers', 'performance-lab' ),
+			);
+			continue;
+		}
+
+		$headers = wp_remote_retrieve_headers( $response );
+		if ( ! is_object( $headers ) && 0 === count( $headers ) ) {
+			// No valid headers retrieved.
+			$final_status   = 'recommended';
+			$fail_details[] = array(
+				'filename' => $filename,
+				'reason'   => __( 'No valid headers retrieved', 'performance-lab' ),
+			);
+			continue;
+		}
+
+		$check = perflab_ffh_check_headers( $headers );
+		if ( isset( $check['passed'] ) && $check['passed'] ) {
+			// This asset passed far-future headers test, no action needed.
+			continue;
+		}
+
+		// If not passed, decide whether to try conditional request.
+		if ( false === $check ) {
```

Comment:
The conditions here seem a bit confusing to me, related to how `perflab_ffh_check_headers()` returns either an `array` or `false`. I think it would be preferable if `perflab_ffh_check_headers` always returned an array. Maybe it could always return an array key for `missing_max_age` which is a boolean. And then this could be changed to:

```suggestion
		$check = perflab_ffh_check_headers( $headers );
		if ( $check['passed'] ) {
			// This asset passed far-future headers test, no action needed.
			continue;
		}

		// If not passed, decide whether to try conditional request.
		if ( $check['missing_max_age'] ) {
```

Reply:
Done in https://github.com/WordPress/performance/pull/1727/commits/12d0b71e7726c9aa2433bd4fee60dae6c37a1b60